src/bms/parse.rs

```diff
@@ -55,29 +50,46 @@ pub struct Bms {
     pub non_command_lines: Vec<String>,
 }
 
+/// Bms Parse Output
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct BmsParseOutput {
+    /// The output Bms.
+    pub bms: Bms,
+    /// Errors
+    pub warnings: Vec<ParseWarning>,
+}
+
 impl Bms {
     /// Parses a token stream into [`Bms`] with a random generator [`Rng`].
-    pub fn from_token_stream(
-        token_stream: &TokenStream,
+    pub fn from_token_stream<'a>(
+        token_iter: &mut std::iter::Peekable<impl Iterator<Item = &'a Token<'a>>>,
         rng: impl Rng,
         mut prompt_handler: impl PromptHandler,
-    ) -> Result<Self> {
-        let continue_tokens = parse_control_flow(token_stream, rng)?;
+    ) -> BmsParseOutput {
+        let (continue_tokens, mut errors) = parse_control_flow(token_iter, rng);
```

Comment:
It seems that the parameter requires users too many characteristics. How about doing to take only `impl Iterator<Item=&'a Token<'a>>` and create a peekable iterator from it?